File: hooks/scripts/cmd/scout_block.go
package cmd

import (
	"encoding/json"
	"os"
	"path/filepath"

	"solon-hooks/internal/config"
	"solon-hooks/internal/logging"
	"solon-hooks/internal/scout"

	"github.com/spf13/cobra"
)

var scoutBlockCmd = &cobra.Command{
	Use:   "scout-block",
	Short: "Handle PreToolUse scout-block (blocks disallowed paths/commands)",
	RunE:  runScoutBlock,
}

func runScoutBlock(cmd *cobra.Command, args []string) error {
	if !config.IsHookEnabled("scout-block") {
		os.Exit(0)
	}

	timer := logging.CreateHookTimer("scout-block")

	var raw json.RawMessage
	if err := json.NewDecoder(os.Stdin).Decode(&raw); err != nil || len(raw) == 0 {
		os.Stderr.WriteString("ERROR: Empty input\n")
		timer.End(logging.LogData{Status: "error", Exit: 2})
		os.Exit(2)
	}

	var data struct {
		ToolName  string                 `json:"tool_name"`
		ToolInput map[string]interface{} `json:"tool_input"`
	}
	if err := json.Unmarshal(raw, &data); err != nil {
		os.Stderr.WriteString("WARN: JSON parse failed, allowing operation\n")
		timer.End(logging.LogData{Status: "ok", Exit: 0})
		os.Exit(0)
	}

	if data.ToolInput == nil {
		os.Stderr.WriteString("WARN: Invalid JSON structure, allowing operation\n")
		timer.End(logging.LogData{Status: "ok", Exit: 0})
		os.Exit(0)
	}

	toolName := data.ToolName
	if toolName == "" {
		toolName = "unknown"
	}

	cwd, _ := os.Getwd()
	claudeDir := filepath.Join(cwd, ".claude")

	result := scout.CheckScoutBlock(toolName, data.ToolInput, scout.CheckOptions{
		ClaudeDir:          claudeDir,
		SlignorePath:       filepath.Join(cwd, ".claude", ".slignore"),
		CheckBroadPatterns: true,
	})

	if result.IsAllowedCommand {
		timer.End(logging.LogData{Tool: toolName, Status: "ok", Exit: 0})
		os.Exit(0)
	}

	if result.Blocked && result.IsBroadPattern {
		msg := scout.FormatBroadPatternError(result.Pattern, result.Suggestions)
		os.Stderr.WriteString(msg)
		timer.End(logging.LogData{Tool: toolName, Status: "block", Exit: 2})
		os.Exit(2)
	}

	if result.Blocked {
		msg := scout.FormatBlockedError(result.Path, result.Pattern, toolName, claudeDir)
		os.Stderr.WriteString(msg)
		timer.End(logging.LogData{Tool: toolName, Status: "block", Exit: 2})
		os.Exit(2)
	}

	timer.End(logging.LogData{Tool: toolName, Status: "ok", Exit: 0})
	os.Exit(0)
	return nil
}

File: hooks/scripts/_archived_ts/hooks/scout-block.ts
// PreToolUse: Block access to directories listed in .slignore + overly broad glob patterns
import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";
import { isHookEnabled } from "../lib/config-loader.ts";
import { createHookTimer } from "../lib/hook-logger.ts";
import { checkScoutBlock } from "../scout-block/scout-checker.ts";
import { formatBlockedError } from "../scout-block/error-formatter.ts";
import { formatBroadPatternError } from "../scout-block/broad-pattern-detector.ts";

if (!isHookEnabled("scout-block")) process.exit(0);

const __dirname = path.dirname(fileURLToPath(import.meta.url));

try {
  const timer = createHookTimer("scout-block");
  const hookInput = fs.readFileSync(0, "utf-8");

  if (!hookInput || hookInput.trim().length === 0) {
    process.stderr.write("ERROR: Empty input\n");
    timer.end({ status: "error", exit: 2 });
    process.exit(2);
  }

  let data: { tool_input?: Record<string, unknown>; tool_name?: string };
  try {
    data = JSON.parse(hookInput);
  } catch {
    process.stderr.write("WARN: JSON parse failed, allowing operation\n");
    timer.end({ status: "ok", exit: 0 });
    process.exit(0);
  }

  if (!data.tool_input || typeof data.tool_input !== "object") {
    process.stderr.write("WARN: Invalid JSON structure, allowing operation\n");
    timer.end({ status: "ok", exit: 0 });
    process.exit(0);
  }

  const toolInput = data.tool_input;
  const toolName = data.tool_name || "unknown";
  // Go up from hooks/scripts/hooks/ → hooks/scripts/ → hooks/ → project root with .slignore
  const claudeDir = path.resolve(__dirname, "..", "..", "..");

  const result = checkScoutBlock({
    toolName,
    toolInput,
    options: {
      claudeDir,
      ckignorePath: path.join(claudeDir, ".slignore"),
      checkBroadPatterns: true,
    },
  });

  if (result.isAllowedCommand) {
    timer.end({ tool: toolName, status: "ok", exit: 0 });
    process.exit(0);
  }

  if (result.blocked && result.isBroadPattern) {
    const errorMsg = formatBroadPatternError({
      blocked: true,
      reason: result.reason,
      pattern: result.pattern,
      suggestions: result.suggestions,
    });
    process.stderr.write(errorMsg);
    timer.end({ tool: toolName, status: "block", exit: 2 });
    process.exit(2);
  }

  if (result.blocked) {
    const errorMsg = formatBlockedError({
      path: result.path!,
      pattern: result.pattern!,
      tool: toolName,
      claudeDir,
    });
    process.stderr.write(errorMsg);
    timer.end({ tool: toolName, status: "block", exit: 2 });
    process.exit(2);
  }

  timer.end({ tool: toolName, status: "ok", exit: 0 });
  process.exit(0);
} catch (err) {
  process.stderr.write(`WARN: Hook error, allowing operation - ${(err as Error).message}\n`);
  process.exit(0);
}
